[WF-564] Table component

blurbyte opened the pull request:
# Table component, and indeterminate Checkbox

## [WF-564](https://datacamp.atlassian.net/browse/564), [WF-627](https://datacamp.atlassian.net/browse/627), [WF-671](https://datacamp.atlassian.net/browse/671)

## Changes

### 🚀 Major Changes

- Added new responsive `Table` component
  - Scrollable when content of the table doesn't fit in the viewport
  - Dark mode
  - Various utility components, which otherwise would be tricky to implement
    - `HeadCell` with sorting
    - `HeadCellCheckbox` and `CellCheckbox`
    - `CellMenu` (with `CellMenuCategory` and `CellMenuItem` re-exported from `Menu` for convenience)
- Added indeterminate `Checkbox` state

**TODO:**
- Use `Table` internally for props
- Add inline props and subcomponents to docs
- Investigate issue in docs when at certain breakpoint (around 850px) it doesn't want to shrink into container (might have something to do with grid layout)

## Screenshot(s):
<img width="819" alt="Screenshot 2022-10-20 at 15 50 09" src="https://user-images.githubusercontent.com/20565536/196968209-cfe71824-27a9-4099-9f54-9a0f98907a37.png">
<img width="459" alt="Screenshot 2022-10-20 at 15 50 20" src="https://user-images.githubuserconten
<img width="817" alt="Screenshot 2022-10-20 at 15 49 56" src="https://user-images.githubusercontent.com/20565536/196968254-b8a358a9-2795-4baa-96a8-d42f786fb4b2.png">
t.com/20565536/196968238-db765006-a7b4-4326-9d96-b5dc5a3618f9.png">
<img width="266" alt="Screenshot 2022-10-20 at 15 50 37" src="https://user-images.githubusercontent.com/20565536/196968265-87722f40-1ad9-487a-bcb2-54ad67bc0e83.png">

ixTec commented on src/checkbox/__tests__/checkbox.e2e.ts:
```diff
@@ -39,4 +48,9 @@ describe('Checkbox', () => {
   });
+
+  it("if indeterminate, when label is clicked doesn't change to checked", () => {
+    cy.loadStory('checkbox-indeterminate');
```
Think this test is incomplete?  Just clicks it but doesn't check for anything after.

I also think we don't want this behaviour? A clicked indeterminate checkbox label should still toggle the checked state because an indeterminate can still have a functional checked/unchecked state (especially useful for forms). See https://baseweb.design/components/checkbox/ or https://ant.design/components/checkbox/ for good examples of this

blurbyte replied:
I'm pretty sure indeterminate takes precedence over checked. You can check in advanced `Table` example how it works.

blurbyte replied:
It is controlled component and doesn't contain any state / logic inside. Not really sure if there is any universal rule whether to show / hide this indeterminate checkmark. Would be great to see some possible implementation example, since I not really sure what are we trying to achieve by adding some extra logic / quasi state internally.

ixTec replied:
I'm definitely not trying to suggest we add some internal logic for updating the state for sure 😅 

It's a tricky situation because the indeterminate state of a checkbox should be largely visual-only. The underlying checkbox still has a checked value.

Right now if I have the following:
```
function Playground() {
  const [checked, setChecked] = useState(false);

  function toggle() {
    setChecked(!checked);
  }

  return (
    <Checkbox
      isIndeterminate
      checked={checked}
      onChange={toggle}
    >
      Checkbox description
    </Checkbox>
  );
}
```
And click the checkbox, the underlying checked value _will change_ to `checked=true`. However the `aria-checked` property will show `indeterminate`. 

There's inconsistency here, and from googling it looks like the checkbox should actually have the property of `indeterminate="true"` instead of the 'checked' property. See https://www.w3.org/TR/html-aria/#att-checked. This property can only be set via js, so will have to be set using a ref.

It also looks like the 'checked' property is calculated automatically from `value` and so we should be using `value` on our input instead of `checked`.

ixTec commented on src/checkbox/__tests__/checkbox.spec.tsx:
```diff
@@ -92,6 +124,30 @@ describe('Checkbox', () => {
+    expect(input).toHaveAttribute('aria-checked', 'mixed');
+  });
+
+  it("renders indeterminate checkbox, even when it's unchecked", () => {
```
Same as previous comment. Perhaps simply renaming these tests to "renders indeterminate checkbox as unchecked/checked" for each of the two tests, although since they aren't rendered as snapshots I think it's bit  odd

blurbyte replied:
just checking here whether checked / unchecked state affects indeterminate (and it shouldn't).

ixTec commented on src/checkbox/__tests__/checkbox.spec.tsx:
```diff
@@ -100,73 +156,97 @@ describe('Checkbox', () => {
-      <Checkbox onChange={jest.fn()}>Taylor Swift</Checkbox>,
-    );
+  describe('renders snapshot', () => {
+    it('with label and input', () => {
```
We should have a snapshot of indeterminate checkbox too

blurbyte replied:
There is snapshot of it at the bottom.

ixTec commented on src/checkbox/styles.ts:
```diff
@@ -70,20 +79,30 @@ const checkmarkBaseStyle = css`
+      : tokens.colors.navyLight;
   }
-  return checked ? tokens.colors.blueDark : tokens.colors.white;
+  return checked || isIndeterminate
```
empty line  after if statement before return

blurbyte commented on src/checkbox/checkbox-internal.tsx:
```diff
@@ -3,50 +3,80 @@ import { mergeProps } from '@react-aria/utils';
+      {children && (
+        <Text
+          as="label"
+          htmlFor={checkboxId}
```
Should we do the same for switch / radio? Plus make it optional (with aria-label required).

ixTec replied:
Yep, sounds like a good idea to me

blurbyte replied:
I will add a story to our board about it.

blurbyte commented on src/checkbox/__tests__/checkbox-indeterminate.story.tsx:
```diff
@@ -0,0 +1,27 @@
+`;
+
+function Story() {
+  const [isChecked, setIsChecked] = useState(false);
```
It's used for checking / unchecking checkbox so it is required.

blurbyte commented on src/checkbox/__tests__/checkbox.spec.tsx:
```diff
@@ -92,6 +124,30 @@ describe('Checkbox', () => {
     expect(input).toBeDisabled();
   });
 
+  it("renders indeterminate checkbox, even when it's checked", () => {
```
It tests whether indeterminate takes precedence over checked state (as it should).

ixTec commented on doc-site/components/markdown-elements.tsx:
```diff
@@ -113,11 +113,26 @@ function CodeBlock({ children }: ContentProps) {
-  width: auto;
+  & th:last-of-type,
+  & td:last-of-type {
+    text-align: right;
```
Hmm I don't know about setting this as the default table styling for markdown tables, since it's only useful for Roadmap. If we add any other tables in `.mdx` down the line, this will have to change. We should probably find a different approach for styling the roadmap table

blurbyte replied:
As long it's the only MDX table not sure if it's worth spending time on it. If we will have more MDX tables happy to update it.

ixTec reviewed (changes requested):
Haven't delved into the implementation of `Table` yet but I have a few comments around docs and examples etc:

- Docs props table doesn't seem to be properly responsive on smaller screen widths (<1200px). Page overflows instead of table scrolling, probably due to `min-width` applied for description
- Looks like best practices not recommended card now has a styling issue from these changes
<img width="1213" alt="Screenshot 2022-10-21 at 11 27 56" src="https://user-images.githubusercontent.com/9938773/197175332-ef4c44a1-8d2d-48ad-8c85-783a0141638e.png">
- Since this is a new component we plan to release, it's important we start releasing them alongside documentation. As a result, we should provide at the very least some initial usage guidelines and best practices for `Table` alongside this, instead of coming back and updating them later.

blurbyte commented on src/dialog/styles.ts:
```diff
@@ -82,10 +79,12 @@ export function bodyStyle({
-      'inset 0 12px 12px -12px rgba(5, 25, 45, 0.3)'}${showShadowTop &&
-      showShadowBottom &&
-      ','}${showShadowBottom && 'inset 0 -12px 12px -12px rgba(5, 25, 45, 0.3)'};
+    box-shadow: ${hasShadowTop &&
```
It's just renaming internal vars, not sure if it's worth documenting.

ixTec replied:
I was initially confused as to what this change was, until I saw the terminology later on the the Table code

blurbyte commented on doc-site/examples/checkbox/states.tsx:
```diff
@@ -12,6 +12,9 @@ function Example() {
       `}
     >
+      <Checkbox isIndeterminate onChange={() => {}}>
+        Indeterminate checkbox
```
Good idea.

blurbyte commented on doc-site/components/best-practices-card.tsx:
```diff
@@ -8,30 +8,49 @@ import {
+    )};
+    border-radius: ${tokens.borderRadius.medium};
+    padding: ${tokens.spacing.medium};
+    padding-top: 20px;
```
Got some style lint error so can't do it that way:
<img width="589" alt="Screenshot 2022-10-21 at 13 56 13" src="https://user-images.githubusercontent.com/20565536/197190283-7420f0ca-0443-4dd8-a9fe-7e072941e82c.png">

blurbyte replied:
Ok, now I see why it wasn't working - I provided 4 values and when last one was moved to next line linter started to complain. Tbh not a big fan of _not_ specifying 2 / 4 in this shorthand notation, had to look up what is happening when you provide only 3.

ixTec replied:
can always specify 4 here, but I think it's good with the three values too

blurbyte commented on doc-site/pages/components/table.mdx:
```diff
@@ -0,0 +1,63 @@
+
+## Usage guidance
+
+<Notification
```
I'm not sure. It's probably most important thing ppl might want to know about this table.

blurbyte commented on doc-site/pages/components/table.mdx:
```diff
@@ -0,0 +1,63 @@
+  title="Customizing column width"
+  description={
+    <>
+      It is possible to fine tune column width by setting <Code>min-width</Code>{' '}
```
It's autoformatted that way with prettier.

ixTec commented on src/table/styles.ts:
```diff
@@ -0,0 +1,305 @@
+    ${!hasShadowLeft &&
+    css`
+      border-top-left-radius: ${tokens.borderRadius.medium};
+      border-bottom-left-radius: ${tokens.borderRadius.medium};
```
can be refactored to `border-radius: ${tokens.borderRadius.medium} 0 0 ${tokens.borderRadius.medium};`

ixTec commented on src/table/styles.ts:
```diff
@@ -0,0 +1,305 @@
+    ${!hasShadowRight &&
+    css`
+      border-top-right-radius: ${tokens.borderRadius.medium};
+      border-bottom-right-radius: ${tokens.borderRadius.medium};
```
can be refactored to `border-radius: 0 ${tokens.borderRadius.medium} ${tokens.borderRadius.medium} 0;`

ixTec commented on src/table/__tests__/table.e2e.ts:
```diff
@@ -0,0 +1,64 @@
+};
+
+describe('Table', () => {
+  describe('when is longer than device width', () => {
```
'...it's longer than...'
or
'...it is longer than...'

ixTec commented on src/table/__tests__/table.e2e.ts:
```diff
@@ -0,0 +1,64 @@
+      cy.get('main').findByLabelText('Regular table').should('exist');
+    });
+
+    it('show shadow scroll hint on both sides, when it was scrolled to the middle', () => {
```
we can scrap 'it was' from these titles of these tests

ixTec commented on src/table/__tests__/table.e2e.ts:
```diff
@@ -0,0 +1,64 @@
+    });
+  });
+
+  // describe('render regular table with all cell types', () => {
```
any reason these are commented out?

blurbyte replied:
Forgot to uncomment these. Was playing with one of the tests above and these were pretty slow.

ixTec commented on doc-site/examples/table/playground-config.ts:
```diff
@@ -0,0 +1,47 @@
+const initialCode = `
+import { Table } from '@datacamp/waffles/table';
+
+function Playground() {
```
I think playground should have at least one of the each cell type for ease of use when playing around component (e.g. a checkbox, menu etc)

blurbyte replied:
I think it is a bit too complex for playground, a lot of unusual handler, etc. which will make it less approachable.

ixTec replied:
This is a fair point, we can keep it as is

ixTec reviewed (changes requested):
Definitely getting closer to having this in top shape I think 💪 

- In hindsight, I think it would've been better to split the changes to Checkbox and Table into two separate PRs, to help with managing the scale of the PR/review.
- Given the complexity of this component, I think it would beneficial to split the directory up into some sub-folders, possibly a 'cells' folder with all cell and cell variant files, or a direct 'subcomponents' folder for all subcomponents instead. It's a minor project management thing but helps keep it a bit more readable when going through the file structure
- Scroll hooks can be refactored for a waffles hook using a similar approach to Tabs
- I agree with splitting up "Advanced Usage" example into individual sortable and checkbox examples

blurbyte commented on src/table/head-cell-checkbox.tsx:
```diff
@@ -0,0 +1,23 @@
+
+  return (
+    <th css={headCellCheckboxStyle()}>
+      <Checkbox {...props} inverted={inverted} aria-label="Select all" />
```
Makes sense.

blurbyte commented on src/table/use-is-scrollable.ts:
```diff
@@ -0,0 +1,34 @@
+import { useState, useEffect } from 'react';
+
+// Check whether wrapper is scrollable or not
+function useIsScrollable(
```
Yes, looks like it's pretty much exactly same functionality. I will move it to helpers, but not sure if it's worth documenting.

ixTec replied:
I think it could be helpful for other teams, doesn't hurt to offer alongside our other things

blurbyte commented on src/table/head-cell.tsx:
```diff
@@ -0,0 +1,60 @@
+          {renderIcon()}
+        </button>
+      ) : (
+        children
```
It's extra HTML wrapper, as Chapeau is not polymorphic. As for icon we might actually need it, since saw something like this in our enterprise app:
<img width="388" alt="Screenshot 2022-10-24 at 10 43 19" src="https://user-images.githubusercontent.com/20565536/197485894-bc9be08f-ac8e-44ce-a5ae-581684845e58.png">

blurbyte commented on src/table/head-cell.tsx:
```diff
@@ -0,0 +1,60 @@
+  /* Show appropriate sort icon, and sets `aria-sort` attribute. */
+  sort?: 'ascending' | 'descending' | 'indeterminate' | 'none';
+  /* Handler called when cell is clicked. */
+  onSort?: React.MouseEventHandler<HTMLButtonElement>;
```
Yes, probably regular `onClick` makes more sense here.

blurbyte commented on doc-site/pages/components/table.mdx:
```diff
@@ -0,0 +1,113 @@
+
+### Table.HeadCell
+
+In addition to all custom props listed below, `Table.HeadCell` accepts all props regular `HTMLTableCell` would, such as `data-` attributes. To customize column width it's possible to provide it via CSS `max-width`, `min-width`, or `width` (in this case percantages are recommended).
```
Wasn't sure if there will be any more in-depth docs in the first place 😄.

blurbyte commented on doc-site/examples/checkbox/indeterminate.tsx:
```diff
@@ -0,0 +1,56 @@
+          setChecked(Array(2).fill(event.target.checked));
+        }}
+      >
+        Check all
```
Probably no need for need for `ScreenreaderOnly`. Label could be updated to be more explicit. Not sure if it needed here since element itself is announced as checkbox.

blurbyte commented:
Since can't comment directly under the comment about indeterminate checkbox:

- Value is doing something different when type is set to checkbox as you can read [here](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox#attr-value).
- All other design systems I checked are handling it in similar way as we do via `aria-checked` attribute
- What is most important, it's announced properly by screenreader

So I don't think we have to do any further refactors of internals.

blurbyte commented on doc-site/pages/components/table.mdx:
```diff
@@ -0,0 +1,113 @@
+  metadata={require('!!extract-react-types-loader!../../../src/table/table-internal')}
+/>
+
+### Table.Head
```
There is a lot going on, but I think we should _always_ keep such subcomponents as separate entries, even if they are identical. It is a matter of improving layout of this section visually. Maybe Maria will have some ideas how to do it. Although it's rather rare case when we have that many subcomponents.

ixTec replied:
I see that as being beneficial yeah, let's speak to Maria to discuss how best we can display this information

blurbyte replied:
But don't think we have to do it right now, after release should be fine.

ixTec commented:
There's some really funky behaviour with different screenreaders and browsers it seems. w3 recommends not to set `aria-checked` when using `type=checkbox` and if it can be different to `checked`, but from testing, not all screenreaders recognise the `indeterminate` attribute for checkboxes... so I am unsure what is best here 😄 

I think we are okay with how you've implemented it, and can always adjust later on if needed

blurbyte commented:
Pushed 1st batch of changes:

- allow setting custom `aria-label` on `HeadCellCheckbox` and `CellCheckbox`, tighten up typings too
- replaced `onSort` with regular `onClick`
- allow setting custom icon next to `HeadCell` label with `xsmall` size injected by default (as in `Tabs`)
- split advanced example into a few smaller ones
- various grammar / naming improvements

As for indeterminate checkbox I imagine a lot of screen readers are scanning only DOM and don't bother with JS-only attrs, so that's it might not work.

So tomorrow I am planning to:

- investigate buggy behavior of table + layout in docs
- refactor scrolling hooks
- add some guidelines

ixTec commented on src/table/__tests__/table.e2e.ts:
```diff
@@ -0,0 +1,64 @@
+      cy.get('table').should('exist');
+    });
+
+    it('show shadow scroll hint to the right, when not scrolled yet', () => {
```
We probably ought to test the mask/gradient in inverted mode too? Just to avoid any accessibility issues with future changes

blurbyte replied:
It is tested already. There are 2 describe blocks at the bottom just for that.

ixTec commented on src/table/__tests__/table.spec.tsx:
```diff
@@ -0,0 +1,518 @@
+
+    it('when sort is not specified, and cell is clicked, appropriate handler is fired', () => {
+      const handleClick = jest.fn();
+      const { container } = render(
```
I would have an empty line before our main render here

blurbyte commented:
Pushed last batch of updates:
- added `useShowScrollHint` hook to handle shadows / masks for `Table`, `Dialog`, and `Tabs`; looks like 2nd ref wasn't even needed
- added basic docs to Table
- removed subcategories from Components in our side navigation, so it's just a flat list now
- some small examples tweaks

blurbyte commented on src/table/head-cell.tsx:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    if (icon) {
+      return icon.props.size ? icon : cloneElement(icon, { size: 'xsmall' });
```
Don't think we can do anything about it. It's more about how some icons look next to text so we would have to SVGs. Don't think there is any universal value we can put here which will fix it for all icons.

blurbyte commented on src/table/head-cell.tsx:
```diff
@@ -0,0 +1,87 @@
+  const { focusProps, isFocusVisible } = useFocusRing();
+  const isSortable = sort !== 'none';
+
+  function renderIcon() {
```
It was switch before. If we want to keep it default case will be rather complex. So pick your poison situation I guess.

ixTec replied:
I think we can clean up to have the `sort === *` comparisons to be a switch, and then keep our `if (icon)` as an if

ixTec commented on doc-site/pages/components/table.mdx:
```diff
@@ -0,0 +1,154 @@
+  <BestPractices.Discouraged>
+
+    - use for layout
+    - when there are fewer sections and each should be ephasized (use a [Card](/components/card))
```
'... emphasized ...'

ixTec commented on doc-site/pages/components/table.mdx:
```diff
@@ -0,0 +1,154 @@
+  <BestPractices.Recommended>
+
+    - use to present tabular data, with a clear, repeating structure
+    - provide `aria-label` for Table
```
I would expand on this and say "provide `aria-label` for context of Table"

ixTec commented on doc-site/pages/components/table.mdx:
```diff
@@ -0,0 +1,154 @@
+
+`Table` helps greatly with comparing, and analyzing data. It help users make informed decsisions. It should be used to present information with a clear, repeating structure.
+
+`Table` which is based on native HTML `table` and it's subcomponents provide a lot of structural information to screen reader users. People who relay on assistive technologies can easily navigate `Table` and identify relationships between data cells and columns. For that reason all of the `Table` components and subcomponents implement HTML `table` elements such as: `table`, `th` or `tr`.
```
"... structural information **_for_** screen reader users."
"... people who _**rely**_ on assistive ..."

ixTec commented on doc-site/pages/components/table.mdx:
```diff
@@ -0,0 +1,154 @@
+- `Table.HeadCell` is used to label columns. Optionally provides sorting functionality. To customize column width it's possible to provide it via CSS `max-width`, `min-width`, or `width` (in this case percantages are recommended).
+- `Table.Cell` is regular data cell.
+
+There are also a few more advanced subcompnents which could be used in some niche situations:
```
**_subcomponents_**

and we can omit "some" as "niche" already implies this

ixTec commented on src/hooks/use-show-scroll-hint.tsx:
```diff
@@ -0,0 +1,65 @@
+ * @param orientation Whether to show hints to the left and right or top and bottom, either `horizontal` or `vertical`
+ * @returns Object of boolean values with `showStartHint`, and `showEndHint` entries
+ */
+function useShowScrollHint(
```
We ought to have some tests for this new hook too

blurbyte replied:
It's kind of tricky, since this hook is all about things you can't test in Jest. So it require spinning Cypress for it. Overall everything it provides is tested in Table and Dialog, so maybe it could be mentioned in comment?

ixTec replied:
I think this is a fair compromise yes

blurbyte replied:
I will add TODO comment here.

ixTec commented on src/tabs/styles.ts:
```diff
@@ -15,13 +15,11 @@ export function tabListStyle({
-        : `linear-gradient(90deg, rgba(0, 0, 0, 1) calc(100% - ${tokens.spacing.xxlarge}), rgba(0, 0, 0, 0))`
-      : 'none'};
+    mask-image: ${isLeftGradientMaskVisible &&
+      `linear-gradient(90deg, rgba(0, 0, 0, 0), rgba(0, 0, 0, 1) ${tokens.spacing.xxlarge}, rgba(0, 0, 0, 1) calc(100% - ${tokens.spacing.xxlarge}), rgba(0, 0, 0, 0))`}${isLeftGradientMaskVisible &&
```
Unsure if it's the styling changes here or the hook changes ,but looks like left gradient isn't being applied correctly for Tabs anymore

![Screenshot 2022-10-26 at 09 43 17](https://user-images.githubusercontent.com/9938773/197978880-67f59858-1b43-4a3a-aed3-90346499c827.png)

blurbyte replied:
Looks like it's styling issue.

ixTec commented on doc-site/pages/components/hooks.mdx:
```diff
@@ -50,6 +50,14 @@ Persists any function and prevents it from being invalidated between renders. Al
 
+## useShowScrollHint
+
+Occasionally for better user experience there is a need to hint that container is scrollable by either showing shadows or griadient masks. Depending on `orientation` (`horizontal` and `vertical` are available) and based on scroll position, show hints at the appropriate positions.
```
_**gradient**_ masks

ixTec reviewed (changes requested):
I think we're very nearly there!

Some minor comments in the code again, and it looks like there are also some leftover comments that need resolving from previous review iterations!

blurbyte commented on doc-site/examples/table/with-checkbox.tsx:
```diff
@@ -0,0 +1,95 @@
+    setData(updatedData);
+  }
+
+  const areChecked = data.map((lang) => {
```
Resulting data structure is, e.g. [true, true, false] with each item representing checked state. Any idea how to better call it?

ixTec replied:
I think `checkedMap` is good here, as it's a map of the checked state of the checkboxes

blurbyte commented on src/table/head-cell.tsx:
```diff
@@ -0,0 +1,87 @@
+} from './styles';
+
+type HeadCellProps = {
+  /* Show appropriate sort icon, and sets `aria-sort` attribute. Actual sorting action should be handled via regular `onClick`. */
```
I think 'icon' here is fine, since it's pretty generic UX pattern to show some kind of indicator column is sortable (be it some arrows, or something more fancy like here).

ixTec replied:
Technically yeah, but given the icons we actually provide I think it's safe to use `Icon` here instead

ixTec commented:
Looks like there's still a small issue with Tabs gradient on scroll, when there's no more overflow on the right:


https://user-images.githubusercontent.com/9938773/198037219-0fcc5b3f-358b-4d8a-b4b2-0a956b11ed6d.mov


We're _super_ close to being ready to merge otherwise. Just this one thing left

ixTec reviewed (approved):
Looks good to merge to me! Great job 🥇
